tsl/src/nodes/decompress_chunk/exec.c

```diff
@@ -500,139 +498,203 @@ perform_vectorized_aggregation(DecompressChunkState *chunk_state)
-									  "bulk decompression",
-									  /* minContextSize = */ 0,
-									  /* initBlockSize = */ 64 * 1024,
-									  /* maxBlockSize = */ 64 * 1024);
-		}
+			// TODO: Handle segment by values
+			CompressedDataHeader *header = (CompressedDataHeader *) PG_DETOAST_DATUM(value);
+			ArrowArray *arrow = NULL;
 
-		DecompressAllFunction decompress_all =
-			tsl_get_decompress_all_function(header->compression_algorithm);
-		Assert(decompress_all != NULL);
+			DecompressAllFunction decompress_all =
+				tsl_get_decompress_all_function(header->compression_algorithm);
+			Assert(decompress_all != NULL);
 
-		MemoryContext context_before_decompression =
-			MemoryContextSwitchTo(chunk_state->bulk_decompression_context);
+			MemoryContext context_before_decompression =
+				MemoryContextSwitchTo(chunk_state->bulk_decompression_context);
 
-		/* Fixme, type for column is wrong because its partial(sum(int4)) */
-		column_description->typid = 23;
+			/* Fixme, type for column is wrong because its partial(sum(int4)) */
+			column_description->typid = 23;
 
-		arrow = decompress_all(PointerGetDatum(header),
-							   column_description->typid,
-							   batch_state->per_batch_context);
+			arrow = decompress_all(PointerGetDatum(header),
+								   column_description->typid,
+								   batch_state->per_batch_context);
 
-		Assert(arrow != NULL);
+			Assert(arrow != NULL);
 
-		MemoryContextReset(chunk_state->bulk_decompression_context);
-		MemoryContextSwitchTo(context_before_decompression);
+			MemoryContextReset(chunk_state->bulk_decompression_context);
+			MemoryContextSwitchTo(context_before_decompression);
 
 #if defined(HAVE__BUILTIN_OP_OVERFLOW) && defined(USE_CUSTOM_SUM_IMPL)
 #define SUM_LANES 4
-		uint32 lane_result[] = { 0, 0, 0, 0 };
-		uint32 lane_buffer[] = { 0, 0, 0, 0 };
-		bool lane_overflow[] = { false, false, false, false };
+			uint32 lane_result[] = { 0, 0, 0, 0 };
+			uint32 lane_buffer[] = { 0, 0, 0, 0 };
+			bool lane_overflow[] = { false, false, false, false };
 
-		int cur_lane = 0;
+			int cur_lane = 0;
 
-		for (int i = 0; i < arrow->length; i++)
-		{
-			const bool arrow_isnull = !arrow_row_is_valid(arrow->buffers[0], i);
-
-			if (!arrow_isnull)
+			for (int i = 0; i < arrow->length; i++)
 			{
-				const uint32 arrow_value = ((uint32 *) arrow->buffers[1])[i];
-				lane_buffer[cur_lane] = arrow_value;
-				cur_lane = (cur_lane + 1) % 4;
-			}
+				const bool arrow_isnull = !arrow_row_is_valid(arrow->buffers[0], i);
 
-			if (cur_lane == 0)
-			{
-				for (int lane_selector = 0; lane_selector < SUM_LANES; lane_selector++)
+				if (!arrow_isnull)
 				{
-					lane_overflow[lane_selector] =
-						__builtin_add_overflow(lane_result[lane_selector],
-											   lane_buffer[lane_selector],
-											   &lane_result[lane_selector]);
+					const uint32 arrow_value = ((uint32 *) arrow->buffers[1])[i];
+					lane_buffer[cur_lane] = arrow_value;
+					cur_lane = (cur_lane + 1) % 4;
 				}
```

Comment:
I wonder if we can get the code generated here that uses the masked add instructions. Maybe you have to add multiple sequential numbers to a lane for that.

Or maybe we can count on the actual values to be zero here, and not use the mask at all, but I'm not sure if that's generally correct, especially if the result comes not from decompression but from other operations.

Reply:
> Or maybe we can count on the actual values to be zero here, and not use the mask at all, but I'm not sure if that's generally correct, especially if the result comes not from decompression but from other operations.

Probably we can't do that, e.g. if we want to use the validity bitmap we got after applying vectorized filters.